drivers/pca9685/pca9685.c

```diff
@@ -0,0 +1,382 @@
+                  __func__, d->params.i2c_dev, dev->params.i2c_addr, ## __VA_ARGS__)
+
+#define EXEC_RET(f) \
+    do { \
+        int _r; \
+        if ((_r = f) != PCA9685_OK) { \
+            DEBUG("[pca9685] %s: error code %d\n", __func__, _r); \
+            return _r; \
+        } \
+    } while(0)
+
+#define EXEC_RET_CODE(f, c) \
+    do { \
+        int _r; \
+        if ((_r = f) != PCA9685_OK) { \
+            DEBUG("[pca9685] %s: error code %d\n", __func__, _r); \
+            return c; \
+        } \
+    } while(0)
+
+#define EXEC(f) \
+    do { \
+        int _r; \
+        if ((_r = f) != PCA9685_OK) { \
+            DEBUG("[pca9685] %s: error code %d\n", __func__, _r); \
+            return; \
+        } \
+    } while(0)
+
+/** Forward declaration of functions for internal use */
+static int _is_available(const pca9685_t *dev);
+static int _init(pca9685_t *dev);
+
+static void _set_reg_bit(uint8_t *byte, uint8_t mask, uint8_t bit);
+
+static int _read(const pca9685_t *dev, uint8_t reg, uint8_t *data, uint32_t len);
+static int _write(const pca9685_t *dev, uint8_t reg, const uint8_t *data, uint32_t len);
+static int _update(const pca9685_t *dev, uint8_t reg, uint8_t mask, uint8_t data);
+
+inline static int _write_word(const pca9685_t *dev, uint8_t reg, uint16_t word);
+
+int pca9685_init(pca9685_t *dev, const pca9685_params_t *params)
+{
+    /* some parameter sanity checks */
+    ASSERT_PARAM(dev != NULL);
+    ASSERT_PARAM(params != NULL);
+    ASSERT_PARAM(params->ext_freq <= 50000000);
+
+    dev->powered_on = false;
+    dev->params = *params;
+
+    DEBUG_DEV("params=%p", dev, params);
+
+    if (dev->params.oe_pin != GPIO_UNDEF) {
+        /* init the pin an disable outputs first */
+        gpio_init(dev->params.oe_pin, GPIO_OUT);
+        gpio_set(dev->params.oe_pin);
+    }
+
+    /* test whether PWM device is available */
+    EXEC_RET(_is_available(dev));
+
+    /* init the PWM device */
+    EXEC_RET(_init(dev));
+
+    return PCA9685_OK;
+}
+
+uint32_t pca9685_pwm_init(pca9685_t *dev, pwm_mode_t mode, uint32_t freq,
+                                                           uint16_t res)
+{
+    /* some parameter sanity checks */
+    ASSERT_PARAM(dev != NULL);
+    ASSERT_PARAM(freq >= 24 && freq <= 1526);
+    ASSERT_PARAM(res >= 2 && res <= 4096);
+    ASSERT_PARAM(mode == PWM_LEFT || mode == PWM_CENTER || mode == PWM_RIGHT);
+
+    DEBUG_DEV("mode=%u freq=%"PRIu32" res=%u", dev, mode, freq, res);
+
+    dev->params.mode = mode;
+    dev->params.freq = freq;
+    dev->params.res = res;
+
+    /* prescale can only be set while in sleep mode (powered off) */
+    if (dev->powered_on) {
+        pca9685_pwm_poweroff(dev);
+    }
+
+    /* prescale = round(clk / (PCA9685_RESOLUTION * freq)) - 1; */
+    float clk = (float)(dev->params.ext_freq ? dev->params.ext_freq
+                                             : PCA9685_OSC_FREQ);
+    float prescale = clk / PCA9685_RESOLUTION / freq;
```

Comment:
Do you really need `float` here?
```C
uint32_t div = PCA9685_RESOLUTION * freq;
uint8_t byte = (clk + div/2) / div - 1;
```

Should achieve the same result (untested)

Reply:
Indeed, `uint32` sholud be enough for a maximum frequency of 1524 Hz and 12 bit resolution. I will change.